scormxblock/static/js/src/scormxblock.js

```diff
@@ -4,105 +4,166 @@ function ScormXBlock(runtime, element, settings) {
-      console.log("LMSGetErrorString");
-      return "Some Error";
+      console.log('LMSGetErrorString');
+      return SCORM_12_ERRORS[errorCode.toString()];
     };
 
     this.LMSGetDiagnostic = function(errorCode) {
-      console.log("LMSGetDiagnostic");
-      return "Some Diagnostice";
+      console.log('LMSGetDiagnostic');
+      return 'Some Diagnostice';
     }
   }
 
   function SCORM_2004_API(){
     this.Initialize = function(){
       console.log('LMSInitialize');
-      return "true";
+      return 'true';
     };
 
     this.Terminate = function() {
-      console.log("LMSFinish");
-      return "true";
+      console.log('LMSFinish');
+      return StoreValues();
     };
 
     this.GetValue = GetValue;
     this.SetValue = SetValue;
 
     this.Commit = function() {
-        console.log("LMSCommit");
-        return "true";
+      console.log('LMSCommit');
+      return StoreValues();
     };
 
     this.GetLastError = function() {
-      console.log("GetLastError");
-      return "0";
+      console.log('GetLastError');
+      return '0';
     };
 
     this.GetErrorString = function(errorCode) {
-      console.log("LMSGetErrorString");
-      return "Some Error";
+      console.log('LMSGetErrorString');
+      return 'Some Error';
     };
 
     this.GetDiagnostic = function(errorCode) {
-      console.log("LMSGetDiagnostic");
-      return "Some Diagnostice";
+      console.log('LMSGetDiagnostic');
+      return 'Some Diagnostice';
     }
   }
 
+  var lastError = 301;
+  /*
+  * These errors has been taken from
+  * https://scorm.com/scorm-explained/technical-scorm/run-time/run-time-reference/
+  */
+  var SCORM_12_ERRORS = {
+    '0': 'No Error',
+    '101': 'General Exception',
+    '201': 'Invalid argument error',
+    '202': 'Element cannot have children',
+    '203': 'Element not an array. Cannot have count',
+    '301': 'Not initialized',
+    '401': 'Not implemented error',
+    '402': 'Invalid set value, element is a keyword',
+    '403': 'Element is read only',
+    '404': 'Element is write only',
+    '405': 'Incorrect Data Type',
+  };
+  var values = [];
+  var LMSValues = {};
+
   var GetValue = function (cmi_element) {
-    var handlerUrl = runtime.handlerUrl(element, 'scorm_get_value');
 
-    var response = $.ajax({
-      type: "POST",
-      url: handlerUrl,
-      data: JSON.stringify({'name': cmi_element}),
-      async: false
-    });
-    response = JSON.parse(response.responseText);
-    console.log("Getvalue for " + cmi_element + " = " + response.value);
-    return response.value
+    if (settings.valid_read_data.includes(cmi_element)){
+      var handlerUrl = runtime.handlerUrl(element, 'scorm_get_values');
+
+      if (values.length===0) {
+        var response = $.ajax({
+          type: 'GET',
+          url: handlerUrl,
+          async: false,
+        });
+        values = JSON.parse(response.responseText);
+      }
+
+      var value = values[cmi_element];
+
+      if (!value) {
```

Comment:
Could you please change the assignation to inline one. @andrey-canon

Reply:
done